http/objectstore/s3api/s3_object_handler.go

```diff
@@ -0,0 +1,402 @@
+	if size > consts.MaxObjectSize {
+		response.WriteErrorResponse(w, r, api_errors.ErrEntityTooLarge)
+		return
+	}
+	cred, _, s3err := s3a.authSys.CheckRequestAuthTypeCredential(context.Background(), r, s3action.PutObjectAction, "testbuckets", "")
+	if s3err != api_errors.ErrNone {
+		response.WriteErrorResponse(w, r, s3err)
+		return
+	}
+	if !s3a.authSys.PolicySys.Head(bucket, cred.AccessKey) {
+		response.WriteErrorResponse(w, r, api_errors.ErrNoSuchBucket)
+		return
+	}
+	dataReader := r.Body
+	hashReader, err1 := hash.NewReader(dataReader, size, clientETag.String(), "", size)
+	if err1 != nil {
+		response.WriteErrorResponse(w, r, api_errors.ErrNewReaderFail)
+		return
+	}
+	defer dataReader.Close()
+	objInfo, err2 := s3a.store.StoreObject(cred.AccessKey, bucket, object, hashReader)
+	if err2 != nil {
+		response.WriteErrorResponse(w, r, api_errors.ErrPutObjectFail)
+		return
+	}
+	setPutObjHeaders(w, objInfo, false)
+	response.WriteSuccessResponseHeadersOnly(w, r)
+}
+
+// GetObjectHandler - GET Object
+// ----------
+// This implementation of the GET operation retrieves object. To use GET,
+// you must have READ access to the object.
+//https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html
+func (s3a *s3ApiServer) GetObjectHandler(w http.ResponseWriter, r *http.Request) {
+	bucket, object := getBucketAndObject(r)
+	var ctx = context.Background()
+	// Check for auth type to return S3 compatible error.
+	// type to return the correct error (NoSuchKey vs AccessDenied)
+	cred, _, s3Error := s3a.authSys.CheckRequestAuthTypeCredential(ctx, r, s3action.GetObjectAction, bucket, object)
+	if s3Error != api_errors.ErrNone {
+		response.WriteErrorResponse(w, r, s3Error)
+		return
+	}
+	if !s3a.authSys.PolicySys.Head(bucket, cred.AccessKey) {
+		response.WriteErrorResponse(w, r, api_errors.ErrNoSuchBucket)
+		return
+	}
+	objInfo, reader, err := s3a.store.GetObject(cred.AccessKey, bucket, object)
+	if err != nil {
+		response.WriteErrorResponseHeadersOnly(w, r, api_errors.ErrGetObjectFail)
+		return
+	}
+	w.Header().Set(consts.AmzServerSideEncryption, consts.AmzEncryptionAES)
+
+	if err = response.SetObjectHeaders(w, r, objInfo); err != nil {
+		response.WriteErrorResponse(w, r, api_errors.ErrSetHeader)
+		return
+	}
+	r1, err := ioutil.ReadAll(reader)
+	if err != nil {
+		response.WriteErrorResponse(w, r, api_errors.ErrNewReaderFail)
+		return
+	}
+	w.Header().Set(consts.ContentLength, strconv.Itoa(len(r1)))
+	response.SetHeadGetRespHeaders(w, r.Form)
+	_, err = w.Write(r1)
+	if err != nil {
+		response.WriteErrorResponse(w, r, api_errors.ErrWriteByteToBodyFail)
+		return
+	}
+}
+
+// HeadObjectHandler - HEAD Object
+//https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html
+// The HEAD operation retrieves metadata from an object without returning the object itself.
+func (s3a *s3ApiServer) HeadObjectHandler(w http.ResponseWriter, r *http.Request) {
+	bucket, object := getBucketAndObject(r)
+	var ctx = context.Background()
```

Comment:
Should avoid to use context.Background(), find a way to use context of the application daemon.

Reply:
”context.Background()“ is temporary and will be changed later